starcluster/streaming_node_add.py

```diff
@@ -0,0 +1,196 @@
+            self.unpropagated_instances, self.instances)
+        if self.unpropagated_instances:
+            log.info("Still waiting for unpropagated instances: "
+                     + str(self.unpropagated_instances))
+        self.instances = self.cluster.get_nodes_or_raise(nodes=self.instances)
+
+    def stream_update_nrm(self):
+        for instance in self.instances:
+            if instance.id not in self.instances_nrm:
+                nrm_cls = partial(NodeRecoveryManager,
+                                  reboot_interval=self.reboot_interval,
+                                  n_reboot_restart=self.n_reboot_restart)
+                if isinstance(instance, Node):
+                    nrm = nrm_cls(instance)
+                else:
+                    nrm = nrm_cls(Node(instance, self.cluster.key_location))
+                self.instances_nrm[instance.id] = nrm
+
+    def stream_instances(self):
+        if not self.instances:
+            return
+
+        ssh_up = self.cluster.pool.map(lambda i: i.is_up(), self.instances)
+        zip_instances = utils.filter_move(
+            lambda i: i[0].state != 'running' or not i[1],
+            zip(self.instances, ssh_up), self.ready_instances,
+            lambda i: i[0])
+        self.instances = [i[0] for i in zip_instances]
+        if self.instances:
+            log.info("Still waiting for instances: " + str(self.instances))
+
+    def stream_manage_reboots(self):
+        dead_instances = []
+        self.instances = \
+            utils.filter_move(lambda i: self.instances_nrm[i.id].check(),
+                              self.instances, dead_instances)
+        for instance in dead_instances:
+            del self.instances_nrm[instance.id]
+
+    def stream_ready_instances(self):
+        for ready_instance in self.ready_instances:
+            log.info("Adding node: %s" % ready_instance.alias)
+            up_nodes = filter(lambda n: n.is_up(), self.cluster.nodes)
+            try:
+                self.cluster.run_plugins(method_name="on_add_node",
+                                         node=ready_instance, nodes=up_nodes)
+                # success
+                del self.instances_nrm[ready_instance.id]
+            except:
+                log.error("Failed to add node {}"
+                          .format(ready_instance.alias), exc_info=True)
+                if self.instances_nrm[ready_instance.id].handle_reboot():
+                    # back to not ready list
+                    self.instances.append(ready_instance)
+                else:
+                    # dead, delete
+                    del self.instances_nrm[ready_instance.id]
+
+    def run(self):
+        """
+        As soon as a new node is ready, run the add plugins commands over it.
+        """
+        interval = self.cluster.refresh_interval
+        log.info("Waiting for one of the new nodes to be up "
+                 "(updating every {}s)".format(interval))
+
+        while True:
```

Comment:
Any reason not to write this as `while any([self.unpropagated_spots, self.spots, self.unpropagated_instances, self.instances])`?